### DISCUSSION TITLE
How to connect cPanel Git Version Control to a GitHub private repository?

### DISCUSSION THREAD

**Discussion by lwcorp** [community/community#40790] [New to GitHub] [answered]
Following #40788, I'd like to connect my hosting's cPanel to a GitHub private repository.
I've tried to do so using 5 simple steps but it doesn't work, can you kindly assist?

I have:
1. Created a cPanel key:
![image](https://user-images.githubusercontent.com/1773306/205525756-e6cfd166-c911-40e4-bd55-41ba91b2cd8a.png)
1. Authorized and View/Downloaded it:
![image](https://user-images.githubusercontent.com/1773306/205525921-3b4300c4-d474-4452-8152-2ccc4a0f0a3f.png)
![image](https://user-images.githubusercontent.com/1773306/205525878-5bde0273-6615-4b5a-9307-2ff1e81f9f07.png)
1. Opened the GitHub private repository's setting, pasted the copied key into Deploy Keys (including write access) and hit Add Key:
![image](https://user-images.githubusercontent.com/1773306/205526021-98f22c7f-38b2-42af-89c7-42459c5293c1.png)
1. Opened the GitHub private repository's homepage, clicked Code, chose SSH and copied the result:
![image](https://user-images.githubusercontent.com/1773306/206850614-ee4e540d-ca7a-45e3-8e81-83035e365daa.png)
1. Went back to cPanel's Git Version Control, pasted the URL into **Clone URL**...and got an error:

> **Error**: (XID bvp8gj) “/usr/local/cpanel/3rdparty/bin/git” reported error code “128” when it ended: Warning: Permanently added 'github.com,\[IP Address\]' (ECDSA) to the list of known hosts. git@github.com: Permission denied (publickey). fatal: Could not read from remote repository. Please make sure you have the correct access rights and the repository exists. 

![image](https://user-images.githubusercontent.com/1773306/205526568-35d0d3b4-ad6d-4c10-ae9a-fd16fd112c46.png)

**Comment by lwcorp** [comment, marked as answer]
Well, turns out it's **all about step 1**:
1. It **cannot** be done via UI since it must **not** have a password/passphrase, as confirmed [by GitHub](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) and [by cPanel](https://docs.cpanel.net/knowledge-base/web-services/guide-to-git-set-up-access-to-private-repositories/#generate-an-ssh-key).
2. It must **not** be named anything other than the default `id_rsa/id_rsa.pub` (also see [here](https://stackoverflow.com/a/74365447)).
The thing is cPanel does point to using Terminal for this...only to give **incorrect** instructions to rename it, which fails the whole process.

In other words, step 1 must be:
1. Open cPanel's Terminal and type `ssh-keygen -t rsa -b 4096`.
![image](https://user-images.githubusercontent.com/1773306/206851152-f2792268-ecfc-4e94-bed9-0aa11d25a2d3.png)

Note it's not just removing `-f filename`, but there's also no need for `-C "username@example.com"` since the Terminal way adds it automatically anyway.

The rest of the steps remain the same.

**Of course, this limitation of only 1 single filename for the entire GitHub means you can't use Deploy Keys to connect to more than one private repository.**

**Reply by khaleefa222** [reply]
leave a reply

**Reply by saudesaude** [reply]
Olá, pode me auxiliar?

**Reply by robertvan1** [reply]
This worked for me, except I used ECDSA for the encryption keys, as github no longer supports RSA keys.

`ssh-keygen -t ecdsa -b 521`

**Reply by lwcorp** [reply]
Interesting, [GitHub's own instructions](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#generating-a-new-ssh-key) seems to still accept RSA if it's `-b 4096`. How come?

But if anything they recommend `-t ed25519` so may I ask why did you choose `-t ecdsa` and why `521` for `-b`?

**Reply by shababhsiddique** [reply]
> Well, turns out it's **all about step 1**:
> 
> 1. It **cannot** be done via UI since it must **not** have a password/passphrase, as confirmed [by GitHub](https://docs.github.com/en/developers/overview/managing-deploy-keys#deploy-keys) and [by cPanel](https://docs.cpanel.net/knowledge-base/web-services/guide-to-git-set-up-access-to-private-repositories/#generate-an-ssh-key).
> 2. It must **not** be named anything other than the default `id_rsa/id_rsa.pub` (also see [here](https://stackoverflow.com/a/74365447)).
>    The thing is cPanel does point to using Terminal for this...only to give **incorrect** instructions to rename it, which fails the whole process.
> 
> In other words, step 1 must be:
> 
> 1. Open cPanel's Terminal and type `ssh-keygen -t rsa -b 4096`.
>    ![image](https://user-images.githubusercontent.com/1773306/206851152-f2792268-ecfc-4e94-bed9-0aa11d25a2d3.png)
> 
> Note it's not just removing `-f filename`, but there's also no need for `-C "username@example.com"` since the Terminal way adds it automatically anyway.
> 
> The rest of the steps remain the same.
> 
> **Of course, this limitation of only 1 single filename for the entire GitHub means you can't use Deploy Keys to connect to more than one private repository.**

This saved me a alot of hours so update-

the process is as described by lwcorp
but you can use multiple private repo if you give general access from https://github.com/settings/keys > SSH key
adding the key here isntead of in repository deploy section allows full access to all private repo from cpanel.

**Reply by lwcorp** [reply]
> the process is as described by lwcorp

Thanks!

> but you can use multiple private repo if you give general access from https://github.com/settings/keys > SSH key adding the key here isntead of in repository deploy section allows full access to all private repo from cpanel.

The problem is that's good for private usage, not organization one (see [Stackoverflow question](https://stackoverflow.com/questions/49784822/github-organization-deploy-key)).
For comparison, [in BitBucket](https://confluence.atlassian.com/bitbucketserver/ssh-access-keys-for-system-use-776639781.html) "Project admins can add and manage SSH access keys for **a project**. These keys apply to **every repository in the project**."
Of course, you might create a fake username and use it for a shared SSH key, but that doesn't sound right...

**Reply by druzsik** [reply]
@shababhsiddique I wish I could confirm but your suggestion is not bulletproof:
![godaddy-ssh-github](https://user-images.githubusercontent.com/127244098/224868833-79e65aaa-c489-4ac1-95bd-c0484a867766.png)

**Reply by lwcorp** [reply]
That's unfortunate as it did work in December...I'll have to sometime try to beat the new way unless someone else does first and detail it here.

**Reply by shababhsiddique** [reply]
Interesting, i posted after i personally used it. So maybe something changed with cpanel version? I currently have 2 git private repo added and can still pull update from tehre.

**Reply by druzsik** [reply]
Maybe GitHub has changed since then. Security is a hot issue, you know 2FA in focus. I enabeled 2FA, good fun.

Anyway, I gave it up to play with Git Version Control and GitHub. Take a look at [this guideline](https://docs.github.com/en/authentication/troubleshooting-ssh/error-permission-denied-publickey) . 

_$ ssh-add -l -E sha256_ gives an arror: The agent has no identities. 

If you made a new ssh-rsa as @lwcorp suggests, you would [add identity](https://stackoverflow.com/questions/26505980/github-permission-denied-ssh-add-agent-has-no-identities) with _sh-add /path/to/.ssh/id_rsa_. 

You can find an authorized_keys file in _/path/to/.ssh/id_rsa_. You do not have to authorize the key in CPanel-SSH Access, but you will face the error message (The agent has no identities) over and over again if check it out.

So, we are talking about an identity issue with ssh-rsa - I mean. This could be a problem why [GitHub prefers](https://github.blog/2021-09-01-improving-git-protocol-security-github/) ECDSA and ed25519 to ssh-rsa. To honest, I see the conflict between GitHub and CPanel Git Version Control, hope it is just a matter of time.

**Reply by druzsik** [reply]
Just one more thing:

- $ ssh -vT git@github.com
- "debug1: Local version string SSH-2.0-OpenSSH_5.3"  - in my CPanel terminal - that can hurt protocol security
 
No matter what reasonable step you take if (shared) web hosting server runs old OpenSSH...

**Reply by jg2703** [reply]
This process still works for me:

https://docs.cpanel.net/knowledge-base/web-services/guide-to-git-set-up-access-to-private-repositories/

**Reply by pj8912** [reply]
This works for me!
`ssh-keygen -t ecdsa -b 521 -C "YOUR_GITHUB_EMAIL"`

my openssh version
```
$ ssh -V
OpenSSH_5.3p1, OpenSSL 1.0.1e-fips 11 Feb 2013
```

**Reply by heavent8** [reply]
24th august 2023 and still worked for me

**Reply by lwcorp** [reply]
> 24th august 2023 and still worked for me

Do you mean the marked answer?

**Reply by heavent8** [reply]
yeah

**Reply by pj8912** [reply]
[Connect PRIVATE GITHUB repository to CPANEL hosting account](https://gist.github.com/pj8912/20918c9980e496e42a4cf52972c74c74)

**Reply by achawayne3031** [reply]
what if your cpanel don't have a terminal

**Reply by pj8912** [reply]
@achawayne3031  Not sure if it's their in your default create your ssh key section, if not that and you don't have a terminal, you have to contact your hosting provider for activation. Please do some research before buying your hosting package.

**Reply by Prince25** [reply]
If people stumble upon this, there is [a way to connect more than one repository](https://stackoverflow.com/a/73824580/13915206).

**Reply by Dantechdevs** [reply]
Connecting Multiple Private Repositories — A Working Solution
Building on the accepted answer above, here's a tested workaround for those who need to connect more than one private GitHub repository to cPanel, since the id_rsa naming limitation makes that impossible by default.
The trick is to use an SSH config file to map multiple keys to multiple repos.

Step 1 — Generate a separate key for each repository (run in cPanel Terminal):
bashssh-keygen -t rsa -b 4096 -f ~/.ssh/id_rsa_repo1
ssh-keygen -t rsa -b 4096 -f ~/.ssh/id_rsa_repo2

Leave the passphrase empty when prompted — cPanel and GitHub do not support passphrase-protected keys.


Step 2 — Create the SSH config file:
bashnano ~/.ssh/config
Then add the following block:
Host github-repo1
    HostName github.com
    User git
    IdentityFile ~/.ssh/id_rsa_repo1

Host github-repo2
    HostName github.com
    User git
    IdentityFile ~/.ssh/id_rsa_repo2

Step 3 — Add each public key as a Deploy Key in each repository under:
GitHub → Repository → Settings → Deploy Keys → Add Key
To print your public keys run:
bashcat ~/.ssh/id_rsa_repo1.pub
cat ~/.ssh/id_rsa_repo2.pub

Step 4 — Use the custom Host alias as your Clone URL in cPanel Git Version Control:
git@github-repo1:username/repository1
git@github-repo2:username/repository2

This completely bypasses the single id_rsa filename restriction. Tested and working ✅

**Reply by FaroJoaoFaro** [reply]
> what if your cpanel don't have a terminal

I had to figure out a way to do that on some cpanels without terminal access. Here's a guide with a functional approach to it:

https://medium.com/@farojoaofaro/how-to-clone-your-github-repositories-even-the-private-ones-on-cpanel-ab0063bf67ac

**Comment by khaleefa222** [comment]
leave

**Comment by solomon-ochepa** [comment]
To authenticate or clone your GitHub repository via SSH, kindly remove the ".git" at the end of the URL and it will work just fine.

Example: use "git@github.com:username/repository" instead of "git@github.com:username/repository.git"

Please note that an SSH (public) key is required; Repository->Settings->Deploy keys.

@Solomon Oki Ochepa - Oki Technologies Ltd.

**Reply by AfonsoQ11** [reply]
This does not work
